### PR TITLE
astc-encoder: init at 2.5

### PR DESCRIPTION
**Pull request by dasisdormax**
This builds the astc-encoder for the x86_64 architecture and sets
an astcenc symlink for the most compatible sse2 version.

###### Motivation for this change

I use the ASTC format for texture compression in a mobile game I develop with LibGDX, as it can encode alpha values and is supported by hardware on modern iOS and Android devices. This is a command line utility for converting regular images (optionally created by a texture packer) into the ASTC or KTX format.

Usage example:
- Encoding: `astcenc -cl input.png compressed.astc 6x6 -medium`
- Decoding: `astcenc -dl compressed.astc decompressed.png`

Original source on Github: https://github.com/ARM-software/astc-encoder

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -2171,6 +2171,16 @@
     githubId = 4971975;
     name = "Janne Heß";
   };
+  dasisdormax = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:2174]: Please put this into a separate commit with the message ``maintainers: add dasisdormax`` and sort it alphabetically.

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{ lib, stdenv, fetchFromGitHub, cmake }:
+
+stdenv.mkDerivation rec {
+  pname = "astc-encoder";
+  version = "2.5";
+  outputs = [ "out" ];
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:6]:
```suggestion
```

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
+    "-DISA_AVX2=ON" "-DISA_SSE41=ON" "-DISA_SSE2=ON"
+  ];
+
+  # Link binaries into environment and provide 'astcenc' link
+  postInstall = ''
+    mv $out/astcenc $out/bin
+    ln -s astcenc-sse2 $out/bin/astcenc
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:28]:
```suggestion
  '';
  
  outputs = [ "out" ];
```

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  meta = with lib; {
+    inherit (src.meta) homepage;
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:31]:
```suggestion
    homepage = "https://developer.arm.com/solutions/graphics";
```

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  # NEEDHELP: set the correct flags for the target architecture (arm etc.)
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
+    "-DISA_AVX2=ON" "-DISA_SSE41=ON" "-DISA_SSE2=ON"
+  ];
+
+  # Link binaries into environment and provide 'astcenc' link
+  postInstall = ''
+    mv $out/astcenc $out/bin
+    ln -s astcenc-sse2 $out/bin/astcenc
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:27]:
```suggestion
    ln -s $out/bin/astcenc-sse2 $out/bin/astcenc
```

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    owner = "ARM-software";
+    repo = "astc-encoder";
+    rev = "${version}";
+    sha256 = "0ff5jh40w942dz7hmgvznmpa9yhr1j4i9qqj5wy6icm2jb9j4pak";
+  };
+
+  nativeBuildInputs = [ cmake ];
+
+  # Built with CMake flags from https://github.com/ARM-software/astc-encoder/blob/main/Docs/Building.md
+  # NEEDHELP: set the correct flags for the target architecture (arm etc.)
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:18]: Probably needs to be something like https://github.com/NixOS/nixpkgs/blob/7248f89da2dab8bca3a9f4f11d0407d0c8e25e89/pkgs/development/libraries/libpostal/default.nix#L18.

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
+    "-DISA_AVX2=ON" "-DISA_SSE41=ON" "-DISA_SSE2=ON"
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:21]:
```suggestion
    "-DISA_AVX2=ON"
    "-DISA_SSE41=ON"
    "-DISA_SSE2=ON"
```

avx can be detected with ``enableAVX2 ? stdenv.hostPlatform.avx2Support``

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "astc-encoder";
+  version = "2.5";
+  outputs = [ "out" ];
+
+  src = fetchFromGitHub {
+    owner = "ARM-software";
+    repo = "astc-encoder";
+    rev = "${version}";
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:11]:
```suggestion
    rev = version;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by dasisdormax** [comment]
@SuperSandro2000 Thanks for your suggestions. I added the compile settings for ARM and other platforms.

I tested the changes by cross-compiling with `nix-build --arg crossSystem '(import <nixpkgs/lib>).systems.examples.aarch64-multiplatform' -A astc-encoder`. This worked, but I cannot test if the binary works as intended.

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+  # Link binaries into environment and provide 'astcenc' link
+  postInstall = ''
+    mv $out/astcenc $out/bin;
+    ln -s $out/bin/astcenc-none $out/bin/astcenc;
+  '' + lib.optionalString stdenv.hostPlatform.isx86_64 ''
+    ln -sf $out/bin/astcenc-sse2 $out/bin/astcenc;
+  '';
+
+  outputs = [ "out" ];
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:36]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/graphics/astc-encoder/default.nix (hunk 9 above)
Comment (dasisdormax) [reply pkgs/tools/graphics/astc-encoder/default.nix:36]: ah ok, good to know that this is the default.

File: pkgs/tools/graphics/astc-encoder/default.nix (hunk 7 above)
Comment (dasisdormax) [reply pkgs/tools/graphics/astc-encoder/default.nix]: Ahh I just noticed the comment about detecting AVX2 support. I think I can use that to select the correct build flag for the user's machine

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 10]
```diff
@@ -0,0 +1,69 @@
+{ lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  simdExtensions ? null
+}:
```
Comment (SuperSandro2000) [inline pkgs/tools/graphics/astc-encoder/default.nix:6]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, cmake
, simdExtensions ? null
}:
```

**Review by SuperSandro2000 — APPROVED** [review]
LGTM with this change

File: pkgs/tools/graphics/astc-encoder/default.nix [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  ];
+
+  # Link binaries into environment and provide 'astcenc' link
+  postInstall = ''
+    mv $out/astcenc $out/bin
+    ln -s $out/bin/astcenc-${mainBinary} $out/bin/astcenc
+  '';
+
+  meta = with lib; {
+    homepage = "https://developer.arm.com/solutions/graphics";
```
Comment (risicle) [inline pkgs/tools/graphics/astc-encoder/default.nix:54]: I've got to wonder if the github page would be a better homepage. This page is rather broad and even then I wouldn't rely on arm to keep this url stable over any significant period of time.

**Comment by risicle** [comment]
Fails to build for me, macos 10.14.

```
builder for '/nix/store/6a88aqmhd2gr1c4j2nqk9j85n991cb37-astc-encoder-2.5.drv' failed with exit code 2; last 10 log lines:
  [ 61%] Building CXX object Source/CMakeFiles/astcenc-sse2-static.dir/astcenc_platform_isa_detection.cpp.o
  [ 64%] Building CXX object Source/CMakeFiles/astcenc-sse2-static.dir/astcenc_quantization.cpp.o
  [ 67%] Building CXX object Source/CMakeFiles/astcenc-sse2-static.dir/astcenc_symbolic_physical.cpp.o
  [ 70%] Building CXX object Source/CMakeFiles/astcenc-sse2-static.dir/astcenc_weight_align.cpp.o
  [ 73%] Building CXX object Source/CMakeFiles/astcenc-sse2-static.dir/astcenc_weight_quant_xfer_tables.cpp.o
  [ 76%] Linking CXX static library libastcenc-sse2-static.a
  Error running link command: No such file or directory
  make[2]: *** [Source/CMakeFiles/astcenc-sse2-static.dir/build.make:464: Source/libastcenc-sse2-static.a] Error 2
  make[1]: *** [CMakeFiles/Makefile2:115: Source/CMakeFiles/astcenc-sse2-static.dir/all] Error 2
  make: *** [Makefile:171: all] Error 2
```

File: pkgs/tools/graphics/astc-encoder/default.nix (hunk 11 above)
Comment (dasisdormax) [reply pkgs/tools/graphics/astc-encoder/default.nix:54]: I agree that the GitHub page would give more relevant information and documentation for anyone using the package. This page at arm.com is set as homepage in the GitHub project, but if it is not necessary to keep that I can change it.

**Comment by dasisdormax** [comment]
Thanks for noticing the build error on Mac. After installing nix and trying the build myself, I got the same problem. 

I'm not sure if this is easy to fix. If not, I'll mark the package as linux-only in the meantime.

**Comment by dasisdormax** [comment]
@risicle I got the macOS build fixed by switching from stdenv to gccStdenv

**Comment by risicle** [comment]
Ah well done - I'm sure it's possible to get it working using clang but it's a bit beyond my darwin-fu.
